### ISSUE TITLE
Improve detection of config/server.js

### ISSUE THREAD

**Issue by paulovieira** [strapi/strapi#8791]
### Describe the bug

When using emacs to edit `config/server.js` (which would be a natural choice in a production environment), a new file is created by emacs (a backup): `config/server.js~` (note the tilde character at the end).

For some reason strapi will also read this backup file (I confirmed by adding a quick console.log in `lib/core/app-configuration/config.loader.js`).

The problem is that the contents in the actual configuration file (`config/server.js`) now seem to be ignored and strapi will use the default values instead (which are hardcoded in `lib/core/app-configuration/index.js`). 

That is, even if both files are equal, and even if the port is set as `1338`, strapi will use the default `1337` (but this problem is verified for any other property in the configuration).

### Steps to reproduce the behavior

1) `cp config/server.js config/server.js~`
2) change any config property in `config/server.js` (example: `port`)
3) verify that the configuration in `config/server.js` is now completely ignored

### Expected behavior

The `config/server.js~` should not be read by the logic in `lib/core/app-configuration/config.loader.js` (only `config/server.js` should be read)

### System

- Node.js version: 12
- Strapi version: 3.3.4
- Database: sqlite
- Operating system: linux

Thanks for your time!

**Comment by derrickmehaffy** [comment]
While I do agree we should adjust the loading. I don't entirely agree with the root cause (you shouldn't be editing the file in production, it's expected you are using source control)

Likewise we never recommend directly editing the file in production, you should be using environment variables and likewise the .env file to handle changes.

See the following documents as "workarounds" to the bug and how we intend things to be done in production.

https://strapi.io/documentation/v3.x/concepts/configurations.html#environment-variables
https://strapi.io/documentation/v3.x/concepts/configurations.html#environments

**Comment by paulovieira** [comment]
I agree that using an env variable would be more clean. In my case I was just adding the `url`property (following the [nginx guide](https://strapi.io/documentation/v3.x/deployment/nginx-proxy.html)), so something like this would be better:
```js
module.exports = ({ env }) => ({
  ...
  url: env('STRAPI_PUBLIC_URL', '127.0.0.1')
});
```
Apart from this small issue I had no problems in installing and using strapi. Congratulations for this great project. I tried it 2 years ago, it's much more robust now.

**Comment by risseraka** [comment]
Hi there!

Just ran into the issue while editing configuration files in development using emacs.
It took a day to figure out what was going wrong adding debug to follow settings logic.

In `lib/core/app-configuration/config-loader.js`, there is a default case returning an empty object if config file isn't either a `.js` nor `.json` file.

```js
// loadFile()
default:
  return {};
```

Perhaps the issue could be adressed by not returning anything if the file extension isn't supported, likewise:

```js
// loadFile()
const loadFile = file => {
  const ext = path.extname(file);

  switch (ext) {
    case '.js':
      return loadJsFile(file);
    case '.json':
      return loadJSONFile(file);
    default:
      return undefined;
  }
};```

And in the exported function, perhaps consider something like:
```js
const result = loadFile(path.resolve(dir, file.name));

if (result !== undefined) {
  acc[key] = result
}
```

WDYT?

**Comment by derrickmehaffy** [comment]
Fixed on v4 via: https://github.com/strapi/strapi/pull/13336
